### PR TITLE
Filter machines

### PR DESCRIPTION
**Pull request by howbazaar**
Based on unlanded PR #15

### CODE REVIEW

File: controller.go [hunk 1]
```diff
@@ -159,10 +159,27 @@ func (c *controller) Zones() ([]Zone, error) {
 	return result, nil
 }
 
+// MachinesArgs is a argument struct for selecting Machines.
+// Only machines that match the specified criteria are returned.
+type MachinesArgs struct {
+	Hostnames    []string
+	MACAddresses []string
+	SystemIds    []string
```
Comment (mjs) [inline controller.go]: SystemIDs ? (given that you changed this elsewhere)

File: controller.go [hunk 2]
```diff
@@ -259,11 +276,25 @@ func (c *controller) post(path, op string, params url.Values) (interface{}, erro
...
 
+func (c *controller) getQuery(path string, params url.Values) (interface{}, error) {
+	return c._get(path, params)
+}
+
 func (c *controller) get(path string) (interface{}, error) {
+	return c._get(path, nil)
+}
+
+func (c *controller) _get(path string, params url.Values) (interface{}, error) {
```
Comment (mjs) [inline controller.go:287]: Given that `getQuery` and `_get` do the same thing with same signature, why not replace `getQuery` with the existing `_get` and then have `get` call `getQuery`?

File: controller.go [hunk 3]
```diff
@@ -259,11 +276,25 @@ func (c *controller) post(path, op string, params url.Values) (interface{}, erro
...
 	path = EnsureTrailingSlash(path)
 	requestID := nextrequestID()
```
Comment (mjs) [inline controller.go]: nextRequestID?

File: controller_test.go [hunk 4]
```diff
@@ -108,6 +109,34 @@ func (s *controllerSuite) TestMachines(c *gc.C) {
 	c.Assert(machines, gc.HasLen, 3)
 }
 
+func (s *controllerSuite) TestMachinesFilter(c *gc.C) {
+	controller := s.getController(c)
+	machines, err := controller.Machines(MachinesArgs{
+		Hostnames: []string{"untasted-markita"},
```
Comment (mjs) [inline controller_test.go:115]: nice petnames inspired hostname :)

**Comment by mjs** [comment]
Just some little things. LGTM.

File: controller.go (hunk 2 above)
Comment (howbazaar) [inline controller.go:287]: Because it is about to be different. :-)

File: controller.go (hunk 1 above)
Comment (howbazaar) [inline controller.go]: Yes. Good spotting.

File: controller.go (hunk 3 above)
Comment (howbazaar) [inline controller.go]: Yes. Fixed.

File: controller_test.go (hunk 4 above)
Comment (howbazaar) [inline controller_test.go:115]: This is all copied from my current MAAS setup :-)

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**